File: src/main/java/org/moinex/model/investment/InvestmentTarget.java
/*
 * Filename: InvestmentTarget.java
 * Created on: January  2, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.model.investment;

import jakarta.persistence.*;
import java.math.BigDecimal;
import lombok.*;
import org.moinex.model.enums.AssetType;

@Entity
@Table(name = "investment_target")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class InvestmentTarget {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    @Setter(AccessLevel.NONE)
    private Integer id;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false, unique = true)
    private AssetType assetType;

    @Column(name = "target_percentage", nullable = false)
    private BigDecimal targetPercentage;

    @Builder.Default
    @Column(name = "is_active", nullable = false)
    private boolean isActive = true;
}

File: src/main/kotlin/org/moinex/model/investment/InvestmentTarget.kt
/*
 * Filename: InvestmentTarget.kt (original filename: InvestmentTarget.java)
 * Created on: January  2, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 07/03/2026
 */

package org.moinex.model.investment

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import org.moinex.common.toRounded
import org.moinex.model.enums.AssetType
import java.math.BigDecimal

@Entity
@Table(name = "investment_target")
class InvestmentTarget(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    var id: Int? = null,
    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false, unique = true)
    var assetType: AssetType,
    @Column(name = "target_percentage", nullable = false)
    var targetPercentage: BigDecimal,
    @Column(name = "is_active", nullable = false)
    var isActive: Boolean = true,
) {
    init {
        targetPercentage = targetPercentage.toRounded()

        require(targetPercentage >= BigDecimal.ZERO && targetPercentage <= BigDecimal(100)) {
            "Target percentage must be between 0 and 100"
        }
    }

    override fun toString(): String = "InvestmentTarget [id=$id, type=$assetType, target=$targetPercentage%]"
}
